site/app/controllers/pdf/PDFController.php

```diff
@@ -14,22 +14,36 @@ public function __construct(Core $core) {
         parent::__construct($core);
     }
 
-    public function getRealPath(string $file_path, string $id): string {
-        if (str_contains($file_path, "..")) {
-            return "INVALID FILE PATH";
+    /**
+     * returns the path contains the anonymous id given the gradeable id if $real is false
+     * returns the path for the submitted file given the user id if $real is true
+     */
+    public function getPath(string $file_path, string $id, bool $real): string {
+        $path = "";
+        $file_path_parts = explode("/", $file_path);
+
+        if ($real) {
+            if (str_contains($file_path, "..")) {
+                return "INVALID FILE PATH";
+            }
         }
 
-        $real_path = "";
-        $file_path_parts = explode("/", $file_path);
         for ($index = 1; $index < count($file_path_parts); $index++) {
             if ($index === 9) {
-                $real_path .= "/" . $id;
+                if ($real) {
+                    $path .= "/" . $id;
+                }
+                else {
+                    $user_id = $file_path_parts[$index];
+                    $anon_ids = $this->core->getQueries()->getSubmitterIdFromAnonId($user_id, $id);
```

Comment:
This is going to result in repetitive queries which slow down the page load time.  It would be better to move this outside the loop.

Reply:
I think this will only be checked once since its in the condition for index === 9